jetty-io/src/main/java/org/eclipse/jetty/io/DefaultRetainableByteBufferPool.java

```diff
@@ -0,0 +1,390 @@
+    public long getHeapMemory()
+    {
+        return getMemory(false);
+    }
+
+    private long getMemory(boolean direct)
+    {
+        if (direct)
+            return _currentDirectMemory.get();
+        else
+            return _currentHeapMemory.get();
+    }
+
+    @ManagedAttribute("The available bytes retained by direct ByteBuffers")
+    public long getAvailableDirectMemory()
+    {
+        return getAvailableMemory(true);
+    }
+
+    @ManagedAttribute("The available bytes retained by heap ByteBuffers")
+    public long getAvailableHeapMemory()
+    {
+        return getAvailableMemory(false);
+    }
+
+    private long getAvailableMemory(boolean direct)
+    {
+        Pool<RetainableByteBuffer>[] buckets = direct ? _direct : _indirect;
+        long total = 0L;
+        for (int i = 0; i < buckets.length; i++)
+        {
+            Pool<RetainableByteBuffer> bucket = buckets[i];
+            long capacity = (i + 1L) * _factor;
+            total += bucket.values().stream().filter(Pool.Entry::isIdle).count() * capacity;
+        }
+        return total;
+    }
+
+    @ManagedOperation(value = "Clears this RetainableByteBufferPool", impact = "ACTION")
+    public void clear()
+    {
+        clearArray(_direct, _currentDirectMemory);
+        clearArray(_indirect, _currentHeapMemory);
+    }
+
+    private void clearArray(Pool<RetainableByteBuffer>[] poolArray, AtomicLong memoryCounter)
+    {
+        for (Pool<RetainableByteBuffer> retainableByteBufferPool : poolArray)
+        {
+            for (Pool<RetainableByteBuffer>.Entry entry : retainableByteBufferPool.values())
+            {
+                entry.remove();
+                memoryCounter.addAndGet(-entry.getPooled().capacity());
+            }
+        }
+    }
+
+    private void releaseExcessMemory(boolean direct)
+    {
+        long maxMemory = direct ? _maxDirectMemory : _maxHeapMemory;
+        if (maxMemory > 0)
+        {
+            long excess = getMemory(direct) - maxMemory;
+            if (excess > 0)
+                evict(direct, excess);
+        }
+    }
+
+    /**
+     * This eviction mechanism searches for the RetainableByteBuffers that were released the longest time ago.
+     * @param direct true to search in the direct buffers buckets, false to search in the heap buffers buckets.
+     * @param excess the amount of bytes to evict. At least this much will be removed from the buckets.
+     */
+    private void evict(boolean direct, long excess)
+    {
+        if (LOG.isDebugEnabled())
+            LOG.debug("evicting {} bytes from {} pools", excess, (direct ? "direct" : "heap"));
+        long totalMemoryCleared = 0L;
+        long now = System.nanoTime();
+
+        Pool<RetainableByteBuffer>[] buckets = direct ? _direct : _indirect;
+        @SuppressWarnings("unchecked")
+        Pool<RetainableByteBuffer>.Entry[] oldestEntries = new Pool.Entry[buckets.length];
+
+        for (int i = 0; i < buckets.length; i++)
```

Comment:
Should you not iterate from the last, so the larger buffers get evicted first?

Reply:
The whole eviction mechanism was greatly simplified after our discussion. Now a buffer is evicted from every single bucket, so that the iteration order doesn't matter anymore.